### ISSUE TITLE
"Notification" Event Not Firing

### ISSUE THREAD

**Issue by ryanpager** [phonegap/phonegap-plugin-push#158]
So this is the use case im following:

a. Push notification goes to phone
b. User clicks on notification from lock, which opens the app & triggers the 'notification' event
c. Do some obligatory action based on additionalData

This is working perfectly, except it seems like the 'notification' event does not get triggered ever again if the user closes the application through the app launcher (square button & swipe).

So a run down of the problem in short:

a. User opens application, push is registered. 
b. 'Notification' events now get triggered as expected.
c. User closes the application through the app launcher.
d. 'Notification' events no longer get triggered....ever.

Any ideas on this one?

**Comment by fredgalvao** [comment]
Is this on Android?

**Comment by ryanpager** [comment]
Yea, I'm testing this on Android 5.1.1 on a Nexus 5.

I also tried reregistering the device on cold start which fixes the future
notification events (great) but not existing ones already issued.

On Monday, September 21, 2015, Frederico Galvão <notifications@github.com>
wrote:

> Is this on Android?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phonegap/phonegap-plugin-push/issues/158#issuecomment-142124419>
> .
>


-- 
*Ryan Page*
*"In Soviet Russia, Branch Merge You!"*
*w | *http://sudobangbang.io* e |* ryanpager@gmail.com *p |* 469-404-7483

**Comment by macdonst** [comment]
@ryanpager after you force close the app do you still see the new push messages in the notification shade?

**Comment by ryanpager** [comment]
@macdonst Oddly enough yea -- push notifications still work exactly as expected, they just dont trigger the 'notification' event regardless of if the app is in foreground or background.

**Comment by macdonst** [comment]
@ryanpager Okay, that's actually the behavior I would expect. When you close the app through the app launcher you are effectively killing your app so the on('notification') handler has been removed. I'm going to look at waking up the app on incoming notifications in the future but for now this is not handled. 

PR's on this topic are welcome.

**Comment by app2naklabs** [comment]
@ryanpager I am also having similar issue and using phonegap build so can you tell me how you able to receive notification even after force close.

**Comment by ryanpager** [comment]
@app2naklabs I just installed the plugin and registered it like normal. But I am not using phonegap, its a cordova+ionic setup (which i dont think makes too much of a difference). The only different thing im doing is that im re-registering the device on coldstart. The push notification setup is AWS SNS (Amazon) => GCM (Google) => Phone. I havent tried any of this on iPhone yet.

**Comment by app2naklabs** [comment]
Can you please tell me how you are doing that; because when app is completely closed and not in used then in between if any notification is triggered then its not appearing in my device. So If you can tell me what changes you made in .js file that helped you in getting this issue solved. I am using phonegap build for building my app.

**Comment by ryanpager** [comment]
@app2naklabs I dont know how to explain this any better. I am 100% using the stock plugin, with no extra configuration settings, and literally no changes to any file. It just works right out of the box the way I said it does. The only difference between yours and mine is that im using cordova/ionic versus phonegap.

Sample Code:

```coffeescript
settings =
      android:
        senderID: xxxxxxxxxxxxxx
      ios:
        alert: yes
        badge: yes
        sound: yes

# Create instance
pushInstance = PushNotification.init(settings)

# Handle different events
pushInstance.on 'registration', _onRegistration
pushInstance.on 'notification', _onNotification
pushInstance.on 'error', _onError
```

**Comment by Mojo90** [comment]
From my point of view:
Using also cordova.
On Android everything works fine! Even when App is closed and than cold started through a notification the on notification is triggered. And of course Push Notifications are receveid at anytime. (As known by other Threads I am using Parse.com as Backend.)
BUT on iOS the cold start does unfortunately not trigger the on notification. I am very interested though that also on iOS it will be triggered. If iOS App is in Background everything just works fine as on Android but once App is closed and App has to cold start than on notification is not triggering (thus I came to this thread).
I'll have a deeper look at it next week but in case somebody figures this out before pls let me know :).

And yes I am also using Stock Plugin 1.2.3 except I did some customizing on Android for Parse.com which can be read here: #147 (but I think this customizing has nothing to do with working on notification trigger on Android cold start)

**Comment by kunalnagar** [comment]
@ryanpager Facing a similar issue. The ```.on('notification')``` handler is not invoked if the app has been closed from the multi-tasking pane on Android. Instead, it starts off with the device ready code and functions normally as if it was a normal app start.

**Comment by EduardMalik** [comment]
+1
Android 5.1
`.on('notification')` not fired when application closed

**Comment by streetkid** [comment]
Using ionic here. 

Experiencing similar issue when quitting the app in ionic regardless of whether i quit through task manager or through the program close. 

I have a login in my app so it needs to check auth first before redirecting to the relevant page based on the notification that was clicked. The actual handling of notificationReceived is in the controller after login. 
I'll try and move it to before login and see if it fires. 

Does register/init need to happen before notificationReceived, if the device was already registered in a previous session?

**Comment by uriva** [comment]
++

**Comment by jianhao1203** [comment]
Is this enhancement already done?

**Comment by s4lesman** [comment]
Thanks @macdonst.

I get nothing on android (not even a system notification) on a Zenfone 2, Android 5.0 stock rom.

this is the logcat:

```
I/GCM     (19955): GCM message com.test_app.mobile 0:1458748156761978%e60cf8f4f9fd7ecd
W/BroadcastQueue(  673): Reject to launch app com.test_app.mobile/10241 for broadcast: App Op 48
W/GCM-DMM (19955): broadcast intent callback: result=CANCELLED forIntent { act=com.google.android.c2dm.intent.RECEIVE pkg=com.test_app.mobile (has extras) }
```


[https://github.com/phonegap/phonegap-plugin-push/issues/748](url)

**Comment by s4lesman** [comment]
I fixed my issue.. I was sending everything on Data.. that only works when the app is running. I filled the notification object and now it works flawlessly. Thanks @macdonst .

**Comment by superainovacoes** [comment]
+1 for notification triggers when app is closed
also other events like receive/open/dismiss the notification ( more info at #756 )

**Comment by marcoziegaus** [comment]
+1 for notification event in JS when the app is killed/terminated.

I found [this](https://github.com/christocracy/push-cordova-plugin/blob/master/src/android/PushHandlerActivity.java#L34) (restarting the MainActivity if the WebView is not started) somewhere, maybe you could utilise a similar approach?

**Comment by arreumb** [comment]
In native apps notifications are received also when the app is closed, this feature is fundamental.
+1 for receiving notifications when app is closed.

**Comment by macdonst** [comment]
@arreumb you do get the notification in the shade but the JavaScript event is not called until the app starts.

**Comment by arreumb** [comment]
I voted for that feature because I need it but I have a bigger problem: I receive correctly in the plugin registration event the gcm endpoint but when i use it to send a notification from the server I can't get any message in the shade. This is very strange for me because I use the same java code on the server  to send notifications that are received from a javascript service worker and in that case everything works like a charm.

**Comment by NicholasTuck** [comment]
Agreed that this seems very fundamental to notification behavior of all mobile applications.  This functionality does work within Ionic's wrappers of the push notification apis. So it's been solved, but of course we need it within GCM directly, not via Ionic.IO's apis.

If we find the code via ionic we will point you that direction.  
+1 from us.

**Comment by yesnoj** [comment]
actually on iOS the push.on('notification') is called ONLY if the app is in foreground, instead in Android doesn't works.The push arrives in the Notification center (iOS) and in Status Bar (Android).Another a question, how increase the badge if the push.on('notification') is never called with app in background?Has no sense to update the badge number if the app is in foreground.There is such an event which i can handle to increase that value using push.getApplicationIconBadgeNumber() and push.setApplicationIconBadgeNumber()?Thanks

**Comment by marcioibm** [comment]
+1 in many scenarios i must not depend on the user interaction to trigger an event. Javascript should be called even if the app is closed.

**Comment by harley1011** [comment]
+1 definitely need this feature to set badge icons when the app is offline, thanks!

**Comment by macdonst** [comment]
@NicholasTuck oh that's odd since the Ionic wrappers use this plugin. 

@yesnoj & @harley1011 you don't need the app to be running to update the badge, iOS handles it for you. You just need to send the `badge` property in your payload for instance:

```
{
    "aps": {
        "alert": "Setting badge to 3",
        "badge": 3
    }
}
```

**Comment by harley1011** [comment]
@macdonst true, but you can't increment using that method unfortunately so it's not ideal when you are sending out a push notification to a lot of clients because you'll have to either send them all the same badge number or send them each their own badge number which will require you to send out multiple APN calls rather than just sending one message with a list of device ids to APN.

**Comment by macdonst** [comment]
@harley1011 don't increment. Have the app communicate with your app server to tell it when it has read a push message. Keep the canonical count on the server not the device.

**Comment by harley1011** [comment]
@macdonst yeah seems like that's what I'll have to do as I have just found out waking up an iOS device isn't possible with push notifications.

Thanks for your help!

**Comment by macdonst** [comment]
@harley1011 yeah, iOS is a different beast.

**Comment by cjardmore** [comment]
I'm not sure if this will be helpful to anyone, as I'm not a seasoned Java developer.  I've traced this to an empty gCachedExtras variable in PushPlugin.java  Thus the 'extras' data does not have a chance to fire inside the onnotification event.

I did see a comment in this thread stating that (if I understood it correctly) it should be expected behavior for the notification to not fire if the app was not already open.  One specific use case that I think contradicts this would be "deep linking" behavior.  In other words, if I've embedded info about specific page and/or id, etc. that should be opened when the app opens.

In other words:

- app is closed
- I receive notification that says "Something important happened with ID#123"
- I tap the notification
- The extras are populated and relayed and 'onnotification' event is fired so I know what page they need to be sent to and/or any specific ajax requests I need to fire and complete before the page is rendered.

**Comment by yoav-zibin** [comment]
OMG, it took me 4 hours to debug this, and I found the solution (at least the issue that was for me :))

On the server-side, I used both notification and data, i.e., I sent something like this:
{
   "to" : "bk3RNwTe3H0:CI2k_HHwgIpoDKCIZvvDMExUdFQ3P1...",
   "data" : {
     "Nick" : "Mario",
   },
  "notification" : {
      "body" : "great match!",
      "title" : "Portugal vs. Denmark",
    }
 }

(see 
http://stackoverflow.com/questions/35480573/gcmlistenerservice-is-not-called-when-application-is-in-background for more background)

The solution: send instead:
{
   "to" : "bk3RNwTe3H0:CI2k_HHwgIpoDKCIZvvDMExUdFQ3P1...",
   "data" : {
     "Nick" : "Mario",
      "message" : "great match!",
      "title" : "Portugal vs. Denmark",
   },
 }

as is correctly documented in this plugin: https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/PAYLOAD.md   I just stupidly missed it because everything seems to be working fine :)
BUT, when you click on the notification it launches the default activity and not PushHandlerActivity (in fact it doesn't even notify GCMIntentService).
If I look at the logs I see it starts my default activity (which doesn't handle notifications obviously):

05-11 14:46:33.426   971   971 D StatusBar: Clicked on content of 0|com.gamingplatform.tictactoe|0|GCM-Notification:670085550|10107
...
05-11 14:46:33.433   776 16550 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10000000 pkg=com.gamingplatform.tictactoe cmp=com.gamingplatform.tictactoe/.TicTacToe (has extras)} from uid 10107 on display 0
Compare this to when the application is in the foreground, you see it calls GCMIntentService:
05-11 14:29:43.115  1557   577 I GCM     : GCM message com.gamingplatform.tictactoe 0:1462991387314480%224af683224af683
05-11 14:29:43.139  5083  5132 D PushPlugin_GCMIntentService: onMessage - from: 24803504516



For the plugin authors:
You can actually fix this on the client side (my fix was changing the server-side code):
Look at the first solution here:
http://stackoverflow.com/questions/35480573/gcmlistenerservice-is-not-called-when-application-is-in-background

You can extend GcmReceiver:
public class GcmBroadcastReceiver extends GcmReceiver {...

and fix this stupid default behavior (if we set both data and notification).

Not sure how many here have the same issue, but a small fix on the client side can save a lot of hours for many people :)
[mic drop]  :)

**Comment by jedofus** [comment]
hello , how can i get my data from payload of ios for show on alert for example when my app is closed ? it's works for me on Android (foreground,background,app closed) but on Ios it's just work on foreground and background, how can i work this when app closed ?

**Comment by Maximvdw** [comment]
Any update on this or the possible fix @marcoziegaus proposed?

**Comment by ShaunBrassell** [comment]
+1 on launching closed app on push receive (Android)

**Comment by macdonst** [comment]
@Maximvdw the problem with @marcoziegaus suggestion is that forcing the reload of the main activity brings the app to the foreground again which is not exactly what we want. The way Android works is there is no good way to start an Activity in the background.

**Comment by ShaunBrassell** [comment]
For my application I actually want the app to launch ... I will attempt to give @marcoziegaus suggestion a try

**Comment by Maximvdw** [comment]
@ShaunBrassell I am fairly sure that doesn't require any additional changes. Currently my approach opens the app (what is personally for me not what I want). Make sure foreground: true when you send the GCM push message

**Comment by ebarault** [comment]
@macdonst 

> The way Android works is there is no good way to start an Activity in the background.

We use a similar functionnality in the app we're building. Given the push plugin requirements it seems fairly possible to me to wake up a process in the background provided you have another process, part of the overall app that is set up to run forever as a background service . What difficulties do you foresee ?

**Comment by macdonst** [comment]
@ebarault I would love if you had insight on this one. I can deliver the notification payload to the webview when the app is running in the foreground or background. The problem is when a user kills the app via the task manager the activity that instantiates the webview has been killed. Without a webview I can't run the JS code. That is the problem I'm struggling with right now.

**Comment by ebarault** [comment]
@macdonst : i've just read again the whole thread and i feel the user need has drifted. From a user perspective i believe we wan't the onNotification callback to be fired when the app has been killed from the launcher (or any possible reason, including reboot, etc.), right?
Or is this that we want the notifications to be fired on the onNotification callback once the app is relaunched? (which i believe already works for me)

you're right the js onNotifcation handler will have hard time to be fired when the webview got killed (though i'll check if there's not a way to spawn it from some background process). But we could think of using cordova as a bus to communicate the notification payload to some other background service the app developer would implement as a cordova plugin too, and that would handle some service logic based on that payload. It ends up building a blend between an hybrid and a native app I agree. That's the only way we found to build the most advanced service logics.

**Comment by jedofus** [comment]
i solved my problem on ios , i analysed my device log and i see that the function notificationReceived() is called before init function on ios , that's why the callbackid is null. I just recalled notificationsReceived() on didRegisterForRemoteNotificationWithDeviceToken function, and now i can get my data of my notification. #917

**Comment by macdonst** [comment]
I have a hack that works kinda. The GCMIntentService can check to see if the app is running, if not it tells it to start. When the app starts up it checks to see if it should be in the background by looking at an extra in the Intent passed along from when GCMIntentService calls startActivity. If it should be in the background it tells the OS to move itself to the back. It results in a brief flash when the push notification arrives but otherwise works.

The problem is I have no way of injecting the correct code into the onCreate method of the MainActivity.

**Comment by hanicker** [comment]
Adding
launchIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
launchIntent.addFlags(Intent.FLAG_FROM_BACKGROUND);
seems to avoid the flash when starting with the push notification

Anyway I still needed to add 
		Bundle extras = getIntent().getExtras();
		if (extras != null) {
			String startOnBackground = extras.getString("start_on_background");
			if("1".equals(startOnBackground)) {
				moveTaskToBack(true);
			}
		}		
to MainActivity

**Comment by EvsanDlg** [comment]
Guys, should 'notification' events fire for all notifications, that were received before the one that I tapped? Actually, it doesn't matter for me if app coldstarts automaticly after each notification received. But I need when it happens (for instance, if i tap one of notifications received) handlers for other already received messages to trigger at this moment also. But it doesn't.
As a result, data from only tapped notification is handled if app wasn't launched before. This is not expected behavior, is it?

**Comment by Donngal** [comment]
I tried to apply the commit 8b7c972, but not successful. 

The app is still not triggering the Notification Event when started by tap on the Notification. The App starts and the Event does never get Triggered. Perhaps there could be a way to get the "data" of the Notification that triggered the app start, when the deviceready Event is fired?

**Comment by haclairvoyant** [comment]
Hello @macdonst 
Any solution for this #1102 , I had gone through the whole thread but don't see any solution for android.  I am not using "notification" key at all as mentioned in one of the post above but still the same issue.

```
JSONObject jGcmData = new JSONObject();
JSONObject jData = new JSONObject();
jData.put("message", msg);
jData.put("title","My App Name");
jData.put("content-available","1");
jData.put("dataObject",dataObject);
jGcmData.put("to", to);

// What to send in GCM message.
jGcmData.put("data", jData);

Any help would be appreciated.
```

**Comment by HariharanIT** [comment]
put this code : push.on('notification', function(data) {})  inside the .run() function..
i have tried with ionic app..its working..even though app closed in background..
My need is to open the specific page of the app on clicking notifications..When app open push.on('notification', function(data) {}) is called so inside this i used $state.go..to open page

**Comment by haclairvoyant** [comment]
@HariharanIT , have you tried with the latest version of Push plugin, please check the versions I mentioned on #1102 . It is not working when closed from active app list and the thing you mentioned is any how required for the notification to execute.

**Comment by hariharanneo** [comment]
@haclairvoyant i am not tried with the latest version..my point is notification event triggered when user click on notification

**Comment by macdonst** [comment]
@haclairvoyant no, there is no solution for this currently. If the user kills you app you will not get the JS event fired. There is one possible way to fix this but it requires a change in cordova-android that I'm trying to get in.

**Comment by murillobarata** [comment]
This issue was solved?

**Comment by ominfowave** [comment]
i have same issue.
 Do you have any solution?

**Comment by ricardokdz** [comment]
I have same issue on iOS.

Exists good alternative for this plugin?

**Comment by davigonz** [comment]
Same issue on Android 6.0 & iOS 8.1.3.

Any solution with Ionic + Cordova? Thank you

**Comment by murillobarata** [comment]
Does anyone know that other plugin can do this type of notification?

**Comment by Muji86** [comment]
If the app is is in the background and swiped away in Android you need to send ("conent-available': 1) in the payload. This will wake up the app and run the notification callback. It will also work once you restart the phone as your app will still be in the background.
If you force kill the app from app manager you will not receive a notification full stop. This is normal android behaviour and it's also true for all apps like WhatsApp.

**Comment by macdonst** [comment]
Okay, if you use cordova-android 6.0.0 or greater and grab the latest from master then you should be able to send `force-start: 1` as detailed here: https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/PAYLOAD.md#application-force-closed

**Comment by serneelsalexis** [comment]
I've seen [Release 1.9.0](https://github.com/phonegap/phonegap-plugin-push/milestone/10) introduced the `force-start`parameter on Android.  
 
I confirm this is still an issue on latest iOS. Notification is not processed when application was "killed". Even combined with `content-available`payload.

**Comment by TrevorPage** [comment]
I don't mean to hijack the topic, but I would like help in understanding whether I actually need to use `force-restart` to mimic the behaviour I would have on a native application. 

On a native Android application, I would expect a couple of important things: (1) After a device restart, it's possible to configure the GCM Service to start up on device boot. This means that the application code can handle the notifications even if the user hasn't started the app from the launcher since boot. (2) If the user backgrounds the app, then even if the associated `Activity`s are killed by the OS, the `Service` will stay, meaning that push notifications are still handled by the application code.

Now, when it comes to an Ionic application using this library, the things I would like to know are:

(1) When the Ionic application is "in the background", presumably the OS may eventually kill it. But, will the push callback still stay around in a Service? Or, do I need to rely on `force-restart` always just in case the app has been killed? As a native Android developer, the use of the term "background" in the lifecycle flowchart for this plugin worries me, because I'm used to expecting backgrounded code to be very short-lived. 

(2) Do I need to also rely on `force-restart` to ensure the callback is called if the user has never started the application since boot, or does the plugin/framework already handle this (perhaps using `android.intent.action.BOOT_COMPLETED`)?

**Comment by VinceOPS** [comment]
@macdonst Hi Simon! Is this new feature `force-start` supposed to also trigger the notification handler? Because it doesn't seem to trigger it, at least in my app. It **does** (re)start the app, but the notification handler isn't triggered. Am I supposed to combine it with `content-available`, or something?

**Comment by TrevorPage** [comment]
@VinceOPS I'm successfully using `force-start` to start the application into the background and I am combining it with `content-available`. Also, my notification intentionally does NOT contain a `message` or `title` property. All of these factors ensure that the library does NOT automatically create a local notification for you, and the handler is always called, whether the application is in foreground, background, or started into background. This is tested by me on Android so far (not iOS yet). 

(As an aside while I'm here, I encourage anyone using force-start to see this issue I raised, which highlights that you must be really careful that your Android really has been properly updated: https://github.com/phonegap/phonegap-plugin-push/issues/1402#issuecomment-262598891)

**Comment by TrevorPage** [comment]
@VinceOPS Another way of putting it, is that even when `force-start` is used, the flowchart here still applies:

https://github.com/phonegap/phonegap-plugin-push/blob/master/docs/PAYLOAD.md

So, `content-available` is needed in the situation where you want the callback to be called when the application is in the background (which is the case when it has been force-started). 

The reason I personally don't have a title or message is because I only want my callback to handle the message; I don't want automatic local notifications from the push library. However these don't affect whether the callback is called (contrary to what I said above). 

Hope that helps :)

**Comment by VinceOPS** [comment]
@TrevorPage: Hi! Thanks for your answer. Unfortunately, I do need to pass a title/message with this push notification... Yup I checked again and I'm pretty sure that my cordova-android is up to date, with the proper MainActivity!

EDIT - Ok, saw your edit in live. The GitHub team is doing things properly right? :P.
I'm going to try right now with content-available and no title/message. But then, how do you manually trigger a local notif? You just **don't**? My use case is simple. The app may have been killed. The user receives a new message. A push notif is sent. I'd like him to be able to open the app directly on the message (that's why I need the notification handle to be triggered).

Thanks again ;)

**Comment by TrevorPage** [comment]
@VinceOPS Just to ensure you noticed it, beware that I've corrected my earlier wrong comment about the title/message. Whether or not these are present do not affect whether the callback is called; only `content-available` affects the callback being called. 

Having a `title` or `message` in the notification causes the push library to automatically create a local notification for you if, and only **if**, the application is not in the foreground. 

I, personally, omit the title and message from the notification because I don't like this behavour. For reasons specific to my application, I always want the notification to be processed by my own code before any local notification is issued. Therefore, I use `$cordovaLocalNotification` to create a local notification, **if** I want to create one. As I'm not using message and title, I put all my application specific data in custom property names.

If for your purposes you are happy for a push notification from the server to always directly cause a local notification based on the message/title when the app isn't in foreground (i.e. it's background or being force-started into background) then you could make use of message and title.

**Comment by VinceOPS** [comment]
Yup, just noticed that `content-available` was helping triggering the notification handler, even with Title/Message, indeed. 
I also noticed that if I click the local notification too quick (before the app actually fully background started, I guess), the handler is not triggered. But this is not a big deal, at the moment.

Thanks again!

**Comment by TrevorPage** [comment]
@serneelsalexis Regarding your observation on iOS, is there any possibility that you have the iOS equivalent to the problem I had here: https://github.com/phonegap/phonegap-plugin-push/issues/1402#issuecomment-262598891

In other words, rebuild the iOS platform using the very latest platform version?

**Comment by fredgalvao** [comment]
@TrevorPage , cordova-ios doesn't have that same feature available yet (start on background). I may be wrong though, it's been a week since I caught up on the dev mailing list.

**Comment by uj** [comment]
I can confirm that force-start feature is working great on Android.  Will this be available in ios?  To be clear, I am referring to the force-start ability for a push notification to start the app when the app is not running (like after a reboot).  It does not matter too much if it opens the app in the foreground or in the background, although opening it in the background would be preferable.

**Comment by midhunadarvin** [comment]
#167
